Add Logging Handler

waprin opened the pull request:
First stab at addressing #1497. 

Starting this with a baby step while we flesh out what we want. So it doesn't worry about fluentd integration, doesn't worry about async transports, doesn't worry about error reporting integration. Just a minimal handler that connects Python logging to Cloud Logging.

jonparrott commented on docs/logging-usage.rst:
```diff
@@ -314,3 +314,32 @@ Delete a sink:
+
+Python Logging Integration:
+
+It's possible to tie the Python `logging` module directly into Google Cloud Logging. To use it, create a :class:`CloudLoggingAPIHandler <gcloud.logging.CloudLoggingAPIHandler` instance from your Logging client.
```
:mod:`logging`, so intersphinx picks it up.

waprin replied:
done

jonparrott commented on docs/logging-usage.rst:
```diff
@@ -314,3 +314,32 @@ Delete a sink:
+    >>> import gcloud.logging # Don't conflict with standard logging
+    >>> from gcloud.logging.handlers import CloudLoggingAPIHandler
+    >>> client = logging.Client()
+    >>> handler = CloudLoggingAPIHandler(client)
```
I'd prefer just `CloudLoggingHandler`.

jonparrott commented on docs/logging-usage.rst:
```diff
@@ -314,3 +314,32 @@ Delete a sink:
+    >>> cloud_logger.addHandler(handler)
+    >>> cloud_logger.error('bad news') # API call
+
+Note that this handler currently only supports a synchronous API call, which means each logging statement that uses this handler will require an API call.
```
Use an admonition:

```
.. note: ...
```

waprin replied:
good call, done

tseaver commented on docs/logging-handlers.rst:
```diff
@@ -0,0 +1,7 @@
+Logger
+======
```
This title, which appears in side-navigation, should give the user a clue that the API is about integration with the Python `logging` module.

waprin replied:
done, went with "Python Logging Module Handler" if that makes sense

tseaver commented on docs/logging-usage.rst:
```diff
@@ -314,3 +314,36 @@ Delete a sink:
+    >>> import logging
+    >>> import gcloud.logging # Don't conflict with standard logging
+    >>> from gcloud.logging.handlers import CloudLoggingHandler
+    >>> client = logging.Client()
```
Should be:

```python
>>> client = gcloud.logging.Client()
```

waprin replied:
done

tseaver commented on docs/logging-usage.rst:
```diff
@@ -314,3 +314,36 @@ Delete a sink:
+    >>> cloud_logger.addHandler(handler)
+    >>> cloud_logger.error('bad news') # API call
+
+.. note: This handler currently only supports a synchronous API call, which means each logging statement that uses this handler will require an API call.
```
Please drop the text of the note to the next line, indent, and wrap.

waprin replied:
done, I had a bunch of mistakes because I wasn't looking more closely at the generated docs, looks closer to what we want now

jonparrott commented:
One concern I thought of is thread-safety. Logging is ostensibly [thread-safe](https://docs.python.org/2/library/logging.html#thread-safety), but I know for sure our client is not.

tseaver commented on gcloud/logging/handlers.py:
```diff
@@ -0,0 +1,102 @@
+    """
+
+    def __init__(self, client):
+        logging.StreamHandler.__init__(self)
```
Why not `super(CloudLoggingHandler, self).__init__()`?

waprin commented:
For thread safety, I think it's fine because the Python handler base class acquires/releases locks around the emit call.

@tseaver do you think this needs a system test added?

tseaver commented:
@waprin a system test would be a good idea.  Figuring how how to tear down the handler, as plugged into the stdlib `logging` module, would be the only hard part, I think.

waprin commented:
Yeah I think that's unfortunately not quite right on the unit tests either. I try to undo it but I see log statements in the tox output I usually don't see. I'll look into fixing it and adding a system test.

waprin commented:
@tseaver Ok I added system tests and I think logging handlers cleanup is now right in both unit and system tests.

I also added a note to CONTRIBUTING that you need to change service account to Owner for logging. This is required to edit Sinks, not thrilled about it but I'm told it works as intended.

The system tests for logging are all waiting for 2 seconds, they were usually failing for me so I bumped it to 5. Unfortunately logging API is especially eventually-consistent so any of these system tests are always going to be at least a little flaky. 

Given there's a bunch more I want to add to improve this handler, as well as more work in this library for Error Reporting, do you think we can start merging all this work into some sort of development branch? I wrote up an internal doc on the game plan, I can share it with you if you like (tseaver at company?). I think the work I have in mind belongs in this repo, but we can discuss, maybe not or maybe you want it as optional dependency or organized into contrib packages.

/cc @filipjs

/cc @steren (no work on error reporting yet, if you'd prefer I can only cc you on PRs relevant to that)

tseaver commented:
@waprin I just set up a [feature branch](https://github.com/GoogleCloudPlatform/gcloud-python/tree/logging-stdlib-handler-feature) for this work.  Can you open a new PR against that branch?

tseaver commented:
Superseded by #1920.
